src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -5372,6 +5376,8 @@ public void ManifestError(string msg, bool runtimeCritical = false)
                 throw new ArgumentException(msg);
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
+        Justification = "Calls to GetFields on types in mapsTab are safe since all their members are preserved")]
```

Comment:
I would reword/clarify this justification. The call to GetFields is getting all of the enum values of an enum type.

Will the ILLinker ever trim unused enum values from an enum? My guess is "no", but it would be good to verify that. If it is "no", then we should say that in the justification.

Similar to the above comment, I would move the suppress message to just the one place that needs to be suppressed.

And then similar to @MichalStrehovsky's suggestion, we could change the use of Reflection with Enum.GetValues and Enum.GetNames.

Reply:
> And then similar to @MichalStrehovsky's suggestion, we could change the use of Reflection with Enum.GetValues and Enum.GetNames.

`Enum.GetValues` creates an array of the enum type. I'm not sure if this works in ReflectionOnly context. EventSource tries to work in ReflectionOnly context for scenarios I don't know much about. I don't know if it's needed here, but I thought I would call that out in case one of the other reviewers knows if that's acceptable. It's definitely preferred because it doesn't need annotations.

Reply:
It seems like `Enum.GetValues` can be used since the types are explicitly checked to be an `enum` before adding to the collection, `mapsTab` and that would not require annotations or suppressions.

Reply:
> It seems like `Enum.GetValues` can be used since the types are explicitly checked to be an `enum` before adding to the collection, `mapsTab` and that would not require annotations or suppressions. 

To make it clear what I'm hinting at:

```csharp
using System;
using System.Reflection;

Type type = new TypeDelegator(typeof(MyEnum));
Console.WriteLine(type.IsEnum);
Console.WriteLine(new string('-', 80));
foreach (var f in type.GetFields())
    Console.WriteLine(f);
Console.WriteLine(new string('-', 80));
foreach (var v in Enum.GetValues(type))
    Console.WriteLine(v);


enum MyEnum { One, Two }
```

This will print:

```
True
--------------------------------------------------------------------------------
Int32 value__
MyEnum One
MyEnum Two
--------------------------------------------------------------------------------
Unhandled exception. System.NotImplementedException: The method or operation is not implemented.
   at System.Type.GetEnumValues()
   at <Program>$.<Main>$(String[] args) in C:\Users\michals\source\repos\NetCoreConsoleApp1\Program.cs:line 14
```

Does this codepath need to operate in such environment?

Reply:
> Does this codepath need to operate in such environment?

I don't _think_ so.... but we should verify. Looking at how this code uses `FieldInfo`:

https://github.com/dotnet/runtime/blob/cb0e8b474a8f26b43566424d98f3014a0b874a71/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L5466-L5468

```C#
using System;
using System.Reflection;

Type type = new TypeDelegator(typeof(MyEnum));
Console.WriteLine(type.IsEnum);
Console.WriteLine(new string('-', 80));
foreach (var f in type.GetFields())
    Console.WriteLine(f.GetRawConstantValue());

enum MyEnum { One, Two }
```

 will print:

```
True
--------------------------------------------------------------------------------
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at System.Reflection.RtFieldInfo.GetRawConstantValue()
   at Submission#7.<<Initialize>>d__0.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.RunSubmissionsAsync[TResult](ImmutableArray`1 precedingExecutors, Func`2 currentExecutor, StrongBox`1 exceptionHolderOpt, Func`2 catchExceptionOpt, CancellationToken cancellationToken)
   at System.Reflection.RtFieldInfo.GetRawConstantValue()
   at Submission#7.<<Initialize>>d__0.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.RunSubmissionsAsync[TResult](ImmutableArray`1 precedingExecutors, Func`2 currentExecutor, StrongBox`1 exceptionHolderOpt, Func`2 catchExceptionOpt, CancellationToken cancellationToken)
```

Reply:
I see, cannot say definitively if this can be ruled out. Sticking to the local function suppression seems safest. Thanks!

Reply:
Unfortunately, the current code has `BindingFlags` and that will work.

```
using System;
using System.Reflection;

Type type = new TypeDelegator(typeof(MyEnum));
Console.WriteLine(type.IsEnum);
Console.WriteLine(new string('-', 80));
foreach (var f in type.GetFields(BindingFlags.DeclaredOnly | BindingFlags.Public | BindingFlags.Static))
    Console.WriteLine(f.GetRawConstantValue());

enum MyEnum { One, Two }
```

will work:
```
True
--------------------------------------------------------------------------------
0
1
```